misk-feature-testing/src/main/kotlin/misk/feature/testing/FakeFeatureFlags.kt

```diff
@@ -77,76 +77,171 @@ class FakeFeatureFlags @Inject constructor(private val moshi : Provider<Moshi>)
   override fun <T : Enum<T>> getEnum(feature: Feature, clazz: Class<T>): T = getEnum(feature, KEY, clazz)
   override fun <T> getJson(feature: Feature, clazz: Class<T>): T = getJson(feature, KEY, clazz)
 
-  private fun get(feature: Feature, key: String): Any? {
+  private fun <V> getOrDefault(
+    feature: Feature,
+    key: String,
+    defaultValue: V,
+    attributes: Attributes = defaultAttributes
+  ): V {
     FeatureFlagValidation.checkValidKey(feature, key)
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides[MapKey(feature)]
-    }
+    @Suppress("unchecked_cast")
+    return (get(feature, key, attributes) ?: defaultValue) as V
   }
 
-  private fun <V> getOrDefault(feature: Feature, key: String, defaultValue: V): V {
+  private fun get(
+    feature: Feature,
+    key: String,
+    attributes: Attributes = defaultAttributes
+  ): Any? {
     FeatureFlagValidation.checkValidKey(feature, key)
-    @Suppress("unchecked_cast")
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides.getOrDefault(MapKey(feature), defaultValue)
-    } as V
+    // Override value lookup is structured in 3 levels: feature, key, and attributes level. The
+    // logic to get the override value is as follows:
+    //
+    // 1. Return feature level override if there is no override at the key level.
+    // 2. if there is an override at the key level, then:
+    //   2.1 Return the override value associated with the attribute override that contains
+    //       all the attributes of the given attributes.
+    //   2.2 If there is no match, return the override value defined at the key level.
+    val overrideMapValues = overrides[MapKey(feature, key)]
+        ?: return overrides[MapKey(feature)]?.let { it.first().value }
+
+    val overrideMapValuesCopy = PriorityQueue(overrideMapValues)
+    var currentMapValue = overrideMapValuesCopy.poll()
+    while (overrideMapValuesCopy.isNotEmpty() || currentMapValue != null) {
```

Comment:
I find this loop hard to understand. I think the `isNotEmpty()` part of the condition can be safely removed?

Reply:
(because it its not empty, then we know currentMapValue will not be null)

Reply:
You are right. `currentMapValue` will never be null at this point. At this point we want to "test" the overrides we have stored against the `attributes` provided. So we have to iterate over all the overrides until we get a match, or until we exhausted all the overrides. Not sure how to express without a `while` loop.

I have updated the while loop in a new commit.